fpdcc/document-search
# Support read-only and read-write non-staff users

## Overview

Create two new groups of users:

- `Read Only` -- users that can only search and view documents
- `Read/Write` -- users that can search and view documents, as well as create/edit/delete them

Use the built-in Django object permissions system to define the permissions for these two groups. Then, update the backend code and templates to properly restrict CRUD views based on these permissions.

In addition, update the login form to use a custom template, and fix up the erroneous redirects for login/logout.

Connects #16, #36.

## Testing Instructions

* Run the management command to create the new groups:

```
docker-compose run --rm app ./manage.py create_permission_groups --verbose
```

* Run `docker-compose up` and head to http://localhost:8000
* If you're already logged in, log out using the link in the navbar; then, confirm that the login page uses a new custom form
    * Try to enter a bogus username/password and confirm you see appropriate errors
    * Login with your credentials and confirm you get taken to the homepage, not the admin interface
* Create two new users in the admin console (http://localhost:8000/admin/):
    * One with the `Read Only` group (note you have to double-click a group to assign it)
    * One with the `Read/Write` group
    * Note that both users should be `active`, but neither should be `staff` or `superuser`!
* Log in as the `Read/Write` user
    * Confirm you see `Add new` in the navbar
    * Confirm you don't see `Admin console` in the footer
        * Try to visit http://localhost:8000/admin/ and confirm you get shown a friendly error
    * Search for a document, visit a detail page, and confirm that you still see the `Edit` button on the page header and can view the `Edit` form
* Log out and log in as the `Read Only` user
    * Confirm you no longer see `Add new` in the navbar
        * Try to navigate to an `Add new` page manually, e.g. http://localhost:8000/books/create/, and confirm you get a `403 Forbidden` error
    * Confirm you don't see `Admin console` in the footer
    * Search for a document, visit a detail page, and confirm that you no longer see the `Edit` button
        * Try to navigate to an `Edit` page manually, e.g. http://localhost:8000/books/update/1, and confirm you get a `403 Forbidden` error
    * Try to navigate to http://localhost:8000/books/delete/1/ and confirm you get a `403 Forbidden` error
